Implement glob-like pattern matching

MVrachev opened the pull request:
Fixes #1505, #1506

**Description of the changes being introduced by the pull request**:

**Implement glob-like pattern matching - 1505**

According to the recently updated version of the specification the shell
style wildcard matching is glob-like (see https://github.com/theupdateframework/specification/pull/174),
and therefore a path separator in a path should not be matched by a
wildcard in the PATHPATTERN.

That's not what happens with `fnmatch.fnmatch()` which doesn't
see "/" separator as a special symbol.
For example: fnmatch.fnmatch("targets/foo.tgz", "*.tgz") will return
True which is not what glob-like implementation will do.

We should make sure that target_path and the `pathpattern` contain the
same number of directories and because each part of the `pathpattern`
could include a glob pattern we should check that `fnmatch.fnmatch()` is true
on each target and `pathpattern` directory fragment separated by "/".

**Avoid lstrip(os.sep) on target paths - 1506**

For targetpath: we don't want to support corner cases such as
file paths starting with separator.
Why this case should be treated especially than any other case where
you have multiple "/" for example "foo//bar/tar.gz"?

For pathpattern: it's recommended that the separator in the pathpattern
should be "/":
see https://theupdateframework.github.io/specification/latest/#targetpath
I believe it could lead to issues for a client implementation if it
supports arbitrary separators - every implementation needs to choose one
and stick with it.
Then, if we decide that "/" is our separator using lstrip on "os.sep" is
wrong, because the os separator from the server could be different that
the one used in the client.

Because of the above arguments, it makes sense to just remove
lstrip on os separators.

Additionally, document in the public API that we only support "/" as a
separator and don't handle corner cases such as leading separators
in either pathpattern or target_filepath.


Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Implement glob-like pattern matching

According to the recently updated version of the specification the shell
style wildcard matching is glob-like (see https://github.com/theupdateframework/specification/pull/174),
and therefore a path separator in a path should not be matched by a
wildcard in the PATHPATTERN.

That's not what happens with `fnmatch.fnmatch()` which doesn't
see "/" separator as a special symbol.
For example: fnmatch.fnmatch("targets/foo.tgz", "*.tgz") will return
True which is not what glob-like implementation will do.

We should make sure that target_path and the pathpattern contain the
same number of directories and because each part of the pathpattern
could include a glob pattern we should check that fnmatch.fnmatch() is
true on each target and pathpattern directory fragment separated by "/".

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

sechkova commented on tuf/ngclient/updater.py:
```diff
@@ -527,8 +544,9 @@ def _visit_child_role(child_role: Dict, target_filepath: str) -> str:
             # Example: "foo.tgz" should match with "/*.tgz".
-            if fnmatch.fnmatch(
-                target_filepath.lstrip(os.sep), child_role_path.lstrip(os.sep)
+            if _is_target_in_pathpattern(
```
The two arguments are swapped.

MVrachev replied:
Thanks, a great catch!

sechkova commented on tuf/ngclient/updater.py:
```diff
@@ -463,6 +463,23 @@ def _preorder_depth_first_walk(self, target_filepath) -> Dict:
 
+def _is_target_in_pathpattern(targetname, pathpattern):
+    """Determines whether the given "targetname" is inside the "pathpattern"."""
+    if pathpattern == "*":
```
`"*" `should not match path separators too, meaning it matches only files under the root/main/base directory or whatever we should call it.

MVrachev replied:
Yes, now when I think about it you are right.
No special symbol allows you to go recursively through directories if we are speaking for glob-like implementation.

sechkova commented on tuf/ngclient/updater.py:
```diff
@@ -463,6 +463,23 @@ def _preorder_depth_first_walk(self, target_filepath) -> Dict:
+    # Check that pathpattern_dir contains the target_dir and `*` is used at the
+    # end of the pattern.
+    # For example, targetname "foo/bar/zoo/k.tgz" and pattpatern "foo/bar/*".
+    if target_dir.startswith(pathpattern_dir) and pathpattern[-1] == "*":
```
same as above for "*". It should not match '/', right?

MVrachev replied:
Yes, as you have pointed out with * we are considering files only in the current base directory.

sechkova commented on tests/test_updater_ng.py:
```diff
@@ -148,6 +149,29 @@ def test_refresh_with_only_local_root(self):
+    supported_use_cases = [
+      ("foo.tgz", "foo.tgz"),
+      ("foo.tgz", "*"),
+      ("targets/foo.tgz", "*"),
```
This should fail.

MVrachev replied:
Yes, as you have pointed out with * we are considering files only in the current base directory.

sechkova commented on tests/test_updater_ng.py:
```diff
@@ -148,6 +149,29 @@ def test_refresh_with_only_local_root(self):
+      ("foo-version-a.tgz", "foo-version-?.tgz"),
+      ("targets/foo.tgz", "targets/*.tgz"),
+      ("foo/bar/zoo/k.tgz", "foo/bar/zoo/*"),
+      ("foo/bar/zoo/k.tgz", "foo/bar/*")
```
Should fail too.

MVrachev replied:
Yes, as you have pointed out with `*` we are considering files only in the current base directory.

sechkova reviewed (commented):
Comparing the dirnames is a clever trick that matches the most common examples in TUF. But if we want to implement proper "globbing" it should be applied on each of the components of a pathname separately.
Should this be a valid test case too  (I don't know if someone would use it :shrug:) : `("foo/bar/zoo/k.tgz", "foo/*/zoo/*"),`?


Also IIUC, '*'  should not match path separator, I've added comments about it.

MVrachev commented:
After the good review by @sechkova, I decided to change my approach.

As Teodora suggested, I will focus on checking the list of directories returned by `os.path.dirname(pathpattern).split("/")` 
and the list of directories returned by `os.path.dirname(targetname).split("/")`.
If both of them match,  I can use `fnmatch.fnmatch()`.
Also, that way I will be able to accept `*` as a directory name  for pathpattern directories and thus this example:
`targetname = "foo/bar/zoo/k.tgz"` and `pathpattern = foo/*/zoo/*` will be supported.

Please let me know if any of the reviewers have ideas for a combination of `targetname` and `pathpattern` that will return a false positive or false negative

MVrachev commented:
Added type annotations and better function description for `_is_target_in_pathpattern` as suggested by @avelichka.

jku commented:
This seems to be handling directories as a special case. I don't think you'll be able to make that work: as Teodora said globbing should probably be applied on each of the components of a pathname separately (unless of course the python base modules include a method to handle the whole path). 

Think of e.g. these cases to see new issues:
```
("foo/bar", "f?o/bar"),
("foo/bar", "*o/bar"),
```

Other quick comments:
* let's pay attention to os.path usage: both pathpattern and targetpath are (or should be, I think) url fragments and not local filesystem paths. So if we use a path module do we actually want to use `posixpath` and not `os.path`?
* we probably want to do `posixpath.abspath()` at some point to handle cases like "foo//bar" in input -- but I guess that should happen somewhere closer to the API level, and not in this code

MVrachev commented:
> This seems to be handling directories as a special case. I don't think you'll be able to make that work: as Teodora said globbing should probably be applied on each of the components of a pathname separately (unless of course the python base modules include a method to handle the whole path).
> 
> Think of e.g. these cases to see new issues:
> 
> ```
> ("foo/bar", "f?o/bar"),
> ("foo/bar", "*o/bar"),
> ```

I added a new commit in which I change my approach.
In the last commit, I call `fnmatch.fnmatch()` on each directory in the `pathpattern`
separately.
Also, I added two unit test cases testing that:
- `("foo/bar", "f?o/bar")` will pass
- `("foo//bar", "*o/bar")` will fail because of the number of slashes.
I remember we discussed before we don't want to artificially support corner cases.

> 
> Other quick comments:
> 
> * let's pay attention to os.path usage: both pathpattern and targetpath are (or should be, I think) url fragments and not local filesystem paths. So if we use a path module do we actually want to use `posixpath` and not `os.path`?
> * we probably want to do `posixpath.abspath()` at some point to handle cases like "foo//bar" in input -- but I guess that should happen somewhere closer to the API level, and not in this code

We had a discussion with @joshuagl about what targets do we expect as arguments
and his opinion was that `pathpattern` and `targetpath` should point in a local directory
(or at least that's how I interpreted his words).

Additionally, I run all of our tests and checked what do we pass to `fnmatch.fnmatch()` inside 
`_visit_child_role`and `_target_matches_path_pattern()` and none of them had test using URLs.
The old code provides an API call `get_valid_targetinfo()` for the `MultiRepoUpdater`
which seems to me that doesn't handle URLs in any specific way.

jku commented:
> I remember we discussed before we don't want to artificially support corner cases.

If you mean this code expects the "paths" to be canonical representation (so things like "foo//bar" or "foo/../foo/bar" are not supported) that seems correct. Or in other words: I'm fine with requiring well-formed input, I'm cautious about defining things as corner cases :) 

> We had a discussion with @joshuagl about what targets do we expect as arguments
and his opinion was that pathpattern and targetpath should point in a local directory
(or at least that's how I interpreted his words).

The spec is broken about this IMO but I don't think we can or should do that in the implementation -- how could the client know which separator the server is using? The API as it's currently designed should maybe accept filesystem paths as input but I believe the input handling should immediately make the transmogrification to a URL-path so rest of the code can assume everything is URL/posix path... 

Alternatively someone needs to explain to me how we can support random directory separators in all of the code. As an example: if pathpatterns and targetpaths were filesystem paths then you are missing all the windows path tests, and all the tests with half windows paths and half posix paths

jku commented on tuf/ngclient/updater.py:
```diff
@@ -465,30 +465,23 @@ def _preorder_depth_first_walk(self, target_filepath) -> Dict:
-
-    if in_the_same_dir:
-        return fnmatch.fnmatch(targetname, pathpattern)
+    # Every part in the pathpattern could include a regex, that's why
+    # each of the target parts should match the corresponding pathpattern part.
+    for index, pattern_dir in enumerate(pathpattern_parts):
+        if not fnmatch.fnmatch(target_parts[index], pattern_dir):
+            return False
```
this looks a lot better, thanks. You could use `for target_dir, pattern_dir in zip(target_parts, pathpattern_parts)` to avoid the index access.

Also should not call it a regex because it isn't.

MVrachev replied:
Thanks for the `zip` suggestion.
Isn't `f?o` a regex? 
Because that's what could be used as a folder name example: `f?o/bar`.

jku replied:
For TUF it is not a regex, it is a glob pattern

MVrachev replied:
I updated the comments and commit message to use glob pattern instead of regex.

jku commented on tests/test_updater_ng.py:
```diff
@@ -148,6 +149,35 @@ def test_refresh_with_only_local_root(self):
+      ("foo/bar/zoo/k.tgz", "foo/bar/*"),
+      ("foo-version-alpha.tgz", "foo-version-?.tgz"),
+      ("foo//bar", "*o/bar"),
+    ]
```
six cases out of seven actually hit `if len(target_parts) != len(pathpattern_parts)` branch so  fnmatch() only gets called single time in all of this. Please make sure there are failing tests for fnmatch() in the filename part and also in a directory part

MVrachev replied:
I removed two of the test cases that seemed more redundant and added an additional one for a wrong directory regex.
Any other suggestions for tests?

jku commented on tests/test_updater_ng.py:
```diff
@@ -148,6 +149,35 @@ def test_refresh_with_only_local_root(self):
+      ("foo/bar/zoo/k.tgz", "foo/*/zoo/*"),
+      ("foo/bar/zoo/k.tgz", "*/*/*/*"),
+      ("foo/bar", "f?o/bar"),
+    ]
```
please add a case for successful * glob match in a directory part -- in other words a regression test that would have failed the previous version of this PR: `("foo/bar", "*o/bar"),`

MVrachev replied:
I think the test `("foo/bar", "*o/bar")` tests the same use case as line 162:
`("foo/bar/zoo/k.tgz", "*/*/*/*"),`.
Even in line 162 you make sure you use `fnmatch.fnmatch()` on all 4 directories otherwise, it cannot be successful.

jku replied:
I'm just saying: your earlier PR passed all those tests but when I added `"*o/bar"` it did not pass... 

Yes I can see the current implementation should succeed but this is basic regression testing: if you fix a bug, add a test case so the bug won't come back.

MVrachev replied:
I just added that test.

jku reviewed (commented):
Maybe makes sense to finish this after #1463 is in but a few more specific comments:
* the matching looks about right now
* we'll maybe still want to remove the possible starting separators from pathpattern somewhere in this code (I left a longer comment in #1506) -- the separator just shouldn't be clients _os.sep_ because that makes no sense for pathpattern that comes from the server

joshuagl commented:
> We had a discussion with @joshuagl about what targets do we expect as arguments
> and his opinion was that `pathpattern` and `targetpath` should point in a local directory
> (or at least that's how I interpreted his words).

I think I just reiterated what the [spec suggests](https://theupdateframework.github.io/specification/v1.0.20/#targetpath) – that we should expect a relative path using UNIX path separators.

It's possible that we should add an explicit section to the spec, or the oft discussed and frequently desired secondary literature https://github.com/theupdateframework/specification/issues/91, on path handling?

jku reviewed (changes requested):
(marking "changes requested" so it's visible in PR list)

Commit:
Remove lstrip & os.sep on targetpath & pathpattern

For targetpath: we don't want to support corner cases such as
file paths starting with separator.
Why this case should be threated specially than any other case where
you have multiple "/" for example "foo//bar/tar.gz"?

For pathpattern: it's recommended that the separator in the pathpattern
should be "/":
see https://theupdateframework.github.io/specification/latest/#targetpath
I believe it could lead to issues for a client implementation if it
supports arbitrary separators - every implementation needs to choose one
and stick with it.
Then, if we decide that "/" is our separator using lstrip on "os.sep" is
wrong, because the os separator from the server could be different that
the one used in the client.

Because of the above arguments, it makes sense to just remove
lstrip on os separators.

Additionally, document that the target_filepath and the DelegatedRole
paths are expected to be in their canonical forms and only "/" is
supported as target path separator.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

in the public API that we only support "/" as a
separator and don't handle corner cases such as leading separators
in either pathpattern or target_filepath.

MVrachev commented:
I had to rebase on top of develop and move `_is_target_in_pathpattern` inside `tuf/api/metadata.py`.
Additionally, I changed some of the tests and added a commit removing the `lstrip` on `os.sep`

In order to fix #1506 do I need to document that we don't use `lsstrip`, because in the issue description it's said:
`Define a valid target file path and pathpattern strings and reject everything else.`?

MVrachev commented:
I updated the comments and commit message to use `glob pattern` instead of `regex`.

jku reviewed (changes requested):
Thanks, functionality looks fine I think and moving the code to Metadata API looks good to me. Two questions/comments:
* should the function be a stand-alone one? I'm not against this sort of design in general but it looks quite out of place in Metadata API: I think this would be the first top-level function in `tuf.api.metadata` so far. As an alternative it could be a private staticmethod on DelegatedRole, expressing that this is not some generic pattern matching function but an implementation detail of DelegatedRole
* commenting feels heavy (are all comments useful, will they stay up-to-date?): the complete path pattern handling in DelegatedRole is now ~8 lines of fairly straight forward code covered by 13 lines of comments. At least the large comment in `is_delegated_path()` needs an update as it's repetitive and already outdated (as no stripping is done).

Please fix the outdated comment. I'll defer to your decision on the other raised issues.

MVrachev commented:
> Thanks, functionality looks fine I think and moving the code to Metadata API looks good to me. Two questions/comments:
> 
> * should the function be a stand-alone one? I'm not against this sort of design in general but it looks quite out of place in Metadata API: I think this would be the first top-level function in `tuf.api.metadata` so far. As an alternative it could be a private staticmethod on DelegatedRole, expressing that this is not some generic pattern matching function but an implementation detail of DelegatedRole

I agree to make the function a private static method in `DelegatedRole` sounds like the most logical solution.

> * commenting feels heavy (are all comments useful, will they stay up-to-date?): the complete path pattern handling in DelegatedRole is now ~8 lines of fairly straight forward code covered by 13 lines of comments. At least the large comment in `is_delegated_path()` needs an update as it's repetitive and already outdated (as no stripping is done).
> 
> Please fix the outdated comment. I'll defer to your decision on the other raised issues.

I removed the strip comments from `is_delegated_path` as well as another comment from `_is_target_in_pathpattern`  that was non essential I think.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -1098,10 +1097,7 @@ def is_delegated_path(self, target_filepath: str) -> bool:
-                # are also considered matches. Make sure to strip any leading
-                # path separators so that a match is made.
-                # Example: "foo.tgz" should match with "/*.tgz".
+                # pattern (Unix shell-style wildcards).
```
I don't see any code change in this commit, only comments.

MVrachev replied:
Fixed it.

MVrachev commented:
@jku do you think the changes I made in https://github.com/theupdateframework/tuf/pull/1512/commits/6ee2daceb36b40d25ebcda39aa9c855a85320a62 are enough to consider #1506 solved?

jku commented on tuf/api/metadata.py:
```diff
@@ -1057,9 +1056,32 @@ def to_dict(self) -> Dict[str, Any]:
-        the paths that DelegatedRole is trusted to provide"""
+        the paths that DelegatedRole is trusted to provide.
+
+        Note: we only support "/" as a separator and don't handle corner cases
+        such as leading separators in pathpattern or target_filepath. See:
+        https://theupdateframework.github.io/specification/latest/#targetpath
```
documentation should maybe not imply that there are _other unspecified_ "corner cases" that we do not handle.

```suggestion
        target_filepath (and the DelegatedRole paths) are expected to be in 
        their canonical forms, so e.g. "a/b" instead of "a//b" . Only
        "/" is supported as target path separator. Leading separators are not
        handled as special cases (see `TUF specification on targetpath
        https://theupdateframework.github.io/specification/latest/#targetpath`_).
```
Maybe that covers most of what we want to say?

Btw, I'm not sure I got the sphinx link format correct there

MVrachev replied:
I used almost the same text as you suggested, but for the link, I had to read the sphinx docs:
https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages.
Now, the link looks good.

jku reviewed (approved):
LGTM, I left one comment about the docstring but I'll leave decision to you: current one seems ok too.
